### PR TITLE
fix(modal, popover): opening modal and popover now works even if overlay was added to ion-app directly

### PR DESCRIPTION
**Pull request by liamdebeasi**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Issue Number: resolves https://github.com/ionic-team/ionic-framework/issues/23728

## What is the new behavior?

- Modal now opens correctly when placed directly inside of ion-app
- Shadows/backdrop opacities are correctly set for the first open modal
- This changes also fixes a "gotcha" I missed when building declarative modals. If you are creating a modal via `document.createElement`, then the modal is no longer automatically removed from the DOM when the modal is dismissed. Impact on this is low since most developers use `modalController`.

Note: I couldn't think of a better way of doing this fix without the `hasController` internal prop, so if you have other ideas definitely let me know!

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: core/src/components/modal/modal.tsx [hunk 1]
```diff
@@ -328,7 +331,7 @@ export class Modal implements ComponentInterface, OverlayInterface {
      * the overlay is already in the correct place.
```
Comment (amandaesmith3) [inline core/src/components/modal/modal.tsx:331]: This comment should be updated to reflect the new check.

File: core/src/utils/overlays.ts [hunk 2]
```diff
@@ -54,7 +54,7 @@ export const createOverlay = <T extends HTMLIonOverlayElement>(tagName: string,
        * Convert the passed in overlay options into props
        * that get passed down into the new overlay.
        */
-      Object.assign(element, { ...opts });
+      Object.assign(element, { ...opts, hasController: true });
```
Comment (amandaesmith3) [inline core/src/utils/overlays.ts:57]: Does the `hasController` prop need to be added to other overlay components besides `modal` and `popover`? Or is the issue specific to those two? This function gets used in all overlays from what I can tell.

**Review by amandaesmith3 — CHANGES_REQUESTED** [review]
Had a minor change and a question below. I couldn't think of a better way than the `hasController` prop either; since you can theoretically insert a modal anywhere, we can't rely on location in the DOM, and there isn't an inherent difference between modal elements of different types besides the new prop. So I think we're good there.

File: core/src/utils/overlays.ts (hunk 2 above)
Comment (liamdebeasi) [reply core/src/utils/overlays.ts:57]: Not yet - the other overlays can't be used declaratively at the moment. Once they are we should add the prop though.

**Review by amandaesmith3 — APPROVED** [review]

**Pull request merged**